File: 133A/cycle.js
#! /usr/bin/env node
////////////////////////////////////////////////////////
var p = "(acfg)(bcd)(aed)(fade)(bgfae)".split( "" );
const n = p.length;
const back = n - 1;
var tag = new Array( n ).fill( '0' );
////////////////////////////////////////////////////////
function frst_pass() {
    var copy;
    for( let j = 0; j < n; ++j ){
        if( p[j] === '(' ){
            copy = p[j + 1];
            tag[j] = '1';
        }
        if( p[j] === ')' ){
            p[j] = copy;
            tag[j] = '1';
        }
    }
}
////////////////////////////////////////////////////////
function Cycle(){
    this.j = 0; // j: start position
    this.find = function ( cmp, i = this.j + 1 ) {
        // i - starting position
        for(; i < n; ++i ){
            if( cmp( i )) break;
        }
        return i;
    };
    this.main = function () {
        let out = "";
        // find first untagged
        let j = this.find( k => tag[k] === '0' );
        if( j === n ) return out; // no more
        this.j = j; // update for the next cycle
        out = `(${p[j]}`;
        let i = j;
        while( true ){
            tag[i] = '1';
            let key = p[i + 1];
            let cmp = k => p[k] === key;
            i = this.find( cmp, i + 2 );
            if( i === n ){
                if( key === p[j] )break;
                i = this.find( cmp );
                out += key;
            }
        }
        out += ")";
        return out;
    };
}
frst_pass();
var cycle = new Cycle();
var res = ""
while( true ){
    const out = cycle.main();
    if( out === "" )break;
    res += out;
}
console.log( res );
// log: Yeah!

File: 133A/cycle.pl
#!                                             /usr/bin/env perl
################################################################
use strict;
use warnings;
my @f1 = split( //, "(acfg)(bcd)(aed)(fade)(bgfae)="); # formula
my @tg = split( //, "------------------------------"); # tag
my @ut = split( //, "=============================="); # output
my $gp = $#f1; # guard position
################################################################
sub dnp {                       # = = = = = = = = = = = = = = =
    print( join '', @f1, "\n"); #  
    print( join '', @tg, "\n"); # -   -   -   _   -   -   -   _
    print( join '', @ut, "\n"); # o   o   o   O   o   o   o   O
}                               #
################################################################
sub fstpas {                          #
    for( my $j = 0; $j < $gp; $j++) { #
        if( $f1[ $j] eq '(') {        #
            $tg[ $j] = 'v';           #
            $f1[ $gp] = $f1[ $j+1];   # backup next lmnt =======
        } elsif( $f1[ $j] eq ')') {   #
            $tg[ $j] = 'v';           # 
            $f1[ $j] = $f1[ $gp];     # copy ===================
        }                             #
    }                                 #
}                                     # 
################################################################
&dnp();    #                   w   w   w   a       s   t   i   ?
&fstpas(); # _                   o       h   t   i       h   s
&dnp();    #                                                   _
################################################################
my $start = 0; # position of cycle's starting lmnt
my $cx;        # the register( scanning position)
my $bx = 0;    # output position
#######################################################[ start ]
sub find_start() {                   # c       k t   i       t #
    while( $tg[ ++$start] eq "v") {} #     e c         s   u   # 
}                                    #   h         h     o     #
#######################################################[ cycle ]
sub cycle() {                    # |T| | | | | | | | | | |L|E| #
    &find_start();               # | | |à| | | | | | | |¢| | | #
    if( $start eq $gp) {         # | |h| | |s| | | |c| | | | | #
        return 0;                # | | | |‡| | |ª| | |¥| | | | #
    }                            # | | | | | | | | | | | | | | #
##### tag the start ############## | | | |T| | |A| | | | | | | ### 
    $tg[ $start] = "v";          # | |h| | |§| | | |©|¥|©|£|È| #
##### setup cx and the key ####### | | | | | | | | | | | | | | ### 
    $cx = $start + 1;            # |†| | | | | | | | | | | | | #
    $f1[ $gp] = $f1[ $cx];       # | | | | | | | | | | | | | | #
##### output cycle's beginning ### | | |ã| | | | | | | | | | | ###
    $ut[ $bx++] = "(";           # | | | | | | | | | | | | | | #
    $ut[ $bx++] = $f1[ $start];  # |t| |a| |s| | | |c| |c| |e| #
    for( ;;) {                   # | |H| |T| | |A| | |Y| |L| | #
        &scan();                 ###############################
        # ck if key equal to starting lmnt #####################
        if( $f1[ $start] eq $f1[ $gp]) { #######################
            last;                ###############################
        } # output key           ###############################
        $ut[ $bx++] = $f1[ $gp]; ###############################
        $cx = $start; # reset to start #########################
    }                            ###############################
    $ut[ $bx++] = ")";           ###############################
    return 1;                    ###############################
}                                ###############################
#################################################[ scan ]#######
sub scan() { # Expects cx and key set up.                    ###
    for( ;;) { #                                              ..  
        while( $f1[ ++$cx] ne $f1[ $gp]) {}                  ### 
        # Here there are equal ck if cx is at guard position ### ###
        if( $cx == $gp) { last; } #              thats break ###
        # Tag lmnt at cx, setup next one as key and continue ###
        $tg[ $cx++] = "v"; # Now_ cx_ is over_ the next_ key ###
        $f1[ $gp] = $f1[ $cx]; # copy key to_ guard position ###
    }                                                        ###
}                                                            ###
while( &cycle()) {}
&dnp();
################################################################
# log:
################################################################
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #
